include/openssl/hmac.h

```diff
@@ -151,6 +151,76 @@ OPENSSL_EXPORT int HMAC_CTX_copy_ex(HMAC_CTX *dest, const HMAC_CTX *src);
 OPENSSL_EXPORT void HMAC_CTX_reset(HMAC_CTX *ctx);
 
 
+// Precomputed key functions
+
+// HMAC_MAX_PRECOMPUTED_KEY_SIZE is the largest precomputed key size, in bytes.
+#define HMAC_MAX_PRECOMPUTED_KEY_SIZE (2 * (EVP_MAX_MD_CHAINING_LENGTH))
+
+// HMAC_set_precomputed_key_export sets the context |ctx| to allow export of the
+// precomputed key using HMAC_get_precomputed_key. On entry, HMAC_CTX must have
+// been initialized via HMAC_Init_*, and neither HMAC_Update nor HMAC_Final
+// must have been called after the last HMAC_Init_ex. It returns one on success
+// and zero on programmer error.
+// After a successful call to HMAC_set_precomputed_key_export, HMAC_Update and
+// HMAC_Final will fail.
+//
+// Note: The main reason for this function is to satisfy FIPS assertion AS09.16,
+// since HMAC_get_precomputed_key returns key material (i.e., a CSP in NIST
+// terminology).
+OPENSSL_EXPORT int HMAC_set_precomputed_key_export(HMAC_CTX *ctx);
+
+// HMAC_get_precomputed_key writes the precomputed key to |out| and sets
+// |*out_len| to the length of the result. On entry, the function
+// HMAC_set_precomputed_key_export must have been called on |ctx|.
+//
+// If |out| is NULL, only |out_len| is set. After such a call,
+// HMAC_get_precomputed_key can directly be called again with a non-null |out|.
+// But HMAC_Update and HMAC_Final will still fail.
+//
+// If |out| is not NULL, |out| must contain at least |out_len| bytes of space
+// (as output by HMAC_get_precomputed_key). An output size of
+// |HMAC_MAX_PRECOMPUTED_KEY_SIZE| will always be large enough. After a
+// successful call to HMAC_get_precomputed_key with a non-NULL |out|, the
+// context can directly be used for computing an HMAC, as if it was set up by
+// HMAC_Init_*. In particular, in particular, HMAC_Update and HMAC_Final can be
+// called. Further calls to HMAC_get_precomputed_key will fail.
```

Comment:
I think "Further calls ..." is not needed because it was mentioned above that it needs to be preceded by `HMAC_set_precomputed_key_export`.